### PR TITLE
fix(common): Don't run preconnect assertion on the server.

### LINKED ISSUE #56207: ERR_INVALID_URL thrown during prerendering ngSrc with priority attribute
### Which @angular/* package(s) are the source of the bug?

common, platform-server

### Is this a regression?

No

### Description

When using Angular to prerender images with the `ngSrc` directive and the `priority` attribute set, an `ERR_INVALID_UR`L error is thrown. Additionally, when this error occurs, the preload tag is not rendered inside the <head> section of the prerendered document.

This issue is specific to the **non-optimized build configuration** - if I set `optimization` to true in development configuration (`angular.json`) the error is no longer being thrown.

You can reproduce it by using the following command in my repro issue: `npm run watch` 

### Please provide a link to a minimal reproduction of the bug

https://github.com/MintMissy/err-invalid-url-repro

### Please provide the exception or error you saw

```true
ERROR TypeError [ERR_INVALID_URL]: Invalid URL
    at __node_internal_captureLargerStackTrace (node:internal/errors:496:5)
    at new NodeError (node:internal/errors:405:5)
    at new URL (node:internal/url:676:13)
    at getUrl (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\common\fesm2022\common.mjs:6043:46)
    at _PreconnectLinkChecker.assertPreconnect (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\common\fesm2022\common.mjs:6567:20)
    at _NgOptimizedImage.ngOnInit (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\common\fesm2022\common.mjs:6964:17)
    at callHookInternal (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:5150:14)
    at callHook (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:5177:13)
    at callHooks (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:5131:17)
    at executeInitAndCheckHooks (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:5081:9)
    at refreshView (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13015:21)
    at detectChangesInView (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13231:9)
    at detectChangesInViewIfAttached (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13191:5)
    at detectChangesInComponent (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13180:5)
    at detectChangesInChildComponents (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13244:9)
    at refreshView (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13054:13)
    at detectChangesInView (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:13231:9)
    at detectChangesInViewWhileDirty (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:12929:9)
    at detectChangesInternal (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:12908:9)
    at detectChangesInViewIfRequired (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32665:5)
    at _ApplicationRef.detectChangesInAttachedViews (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32500:21)
    at _ApplicationRef._tick (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32469:18)
    at _ApplicationRef.tick (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32458:14)
    at _ApplicationRef._loadComponent (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32552:14)
    at _ApplicationRef.bootstrap (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:32440:14)
    at <anonymous> (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:36496:32)
    at _ZoneDelegate.invoke (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:365:28)
    at Object.onInvoke (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:15743:33)
    at _ZoneDelegate.invoke (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:364:34)
    at _ZoneImpl.run (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:111:43)
    at <anonymous> (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:1192:40)
    at _ZoneDelegate.invokeTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:398:33)
    at AsyncStackTaggingZoneSpec.onInvokeTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:15431:28)
    at _ZoneDelegate.invokeTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:397:38)
    at Object.onInvokeTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\@angular\core\fesm2022\core.mjs:15730:33)
    at _ZoneDelegate.invokeTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:397:38)
    at _ZoneImpl.runTask (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:158:47)
    at drainMicroTaskQueue (P:\Xgrow\err-invalid-url-repro\.angular\prerender-root\7732fdf6-f774-4706-b0ac-59d4ca20857d\node_modules\zone.js\fesm2015\zone-node.js:577:35) {
  input: '/public/angular_logo.png',
  code: 'ERR_INVALID_URL'
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 18.0.2
Node: 18.19.1
Package Manager: npm 10.2.4
OS: win32 x64

Angular: 18.0.1
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1800.2
@angular-devkit/build-angular   18.0.2
@angular-devkit/core            18.0.2
@angular-devkit/schematics      18.0.2
@angular/cli                    18.0.2
@angular/ssr                    18.0.2
@schematics/angular             18.0.2
rxjs                            7.8.1
typescript                      5.4.5
zone.js                         0.14.6
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeanMeche**
The `window` global is patched by domino on the server but the value of `window.location.href` isn't a valid base.

Before this change `getUrl()` would throw when running in devmode on the server.

Fixes #56207

### CODE REVIEW

File: packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts [hunk 1]
```diff
@@ -102,7 +104,7 @@ export class PreconnectLinkChecker {
    * @param originalNgSrc ngSrc value
    */
   assertPreconnect(rewrittenSrc: string, originalNgSrc: string): void {
-    if (!this.window) return;
+    if (!this.window || this.isServer) return;
```
Comment (AndrewKushnir) [inline packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:107]: I'm wondering if we should just check if the current platform is server and exit in this case:
```suggestion
    if (this.isServer) return;
```
I think that was the intention of the `window`-based check.

Comment (JeanMeche) [reply packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:107]: I kept it because `this.window ` is typed as `Window | null`

Comment (AndrewKushnir) [reply packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:107]: Thanks for additional information. In this case we can move this check to the constructor:
```
    if (!this.isServer && typeof win !== 'undefined') {
      this.window = win;
    }
```
So we set the window reference only when we are not on the server. What do you think?

Comment (alan-agius4) [reply packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:107]: @JeanMeche, I understand you didn't originally work on this part, but there's an issue with how we're obtaining the base href, and it also highlights a couple of other issues.

1. We're currently using the Window object, which is incorrect. Instead, we should be using the `PlatformLocation` service. This service will return the correct base when running on the server. When an absolute URL is provided, the `PlatformLocation` service handles it correctly.
2. When a relative URL is provided in the server config, we're currently setting the base href to `/`, which is not valid, as you pointed out. This setting lacks a protocol and makes the `href` invalid. https://github.com/angular/angular/blob/101edda0184b458d331163b7e04280c3b8b48bfc/packages/platform-server/src/domino_adapter.ts#L24. 

Maybe the `getURL` function should be updated to use a `base` instead of a Window object:

```ts
export function getUrl(src: string, base: string): URL
```

This approach is essentially equivalent to using `new URL()`, because if `src` is absolute, `new URL(src)` yields the same result as `new URL(src, win.location.href)`.

Comment (AndrewKushnir) [reply packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:107]: @alan-agius4 quick question: the change is in this PR should make this service a noop on the server (only be invoked in a browser). Would you recommend changing the mentioned logic anyway?

Comment (JeanMeche) [reply packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts]: I agree that accessing the window global isn't great. I've refactored to use `PlatformLocation`.

**Review by AndrewKushnir — APPROVED** [review]

File: packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts [hunk 2]
```diff
@@ -42,14 +45,13 @@ export class LCPImageObserver implements OnDestroy {
   // Map of full image URLs -> original `ngSrc` values.
   private images = new Map<string, ObservedImageState>();
 
-  private window: Window | null = null;
   private observer: PerformanceObserver | null = null;
+  private platformLocation = inject(PlatformLocation);
+  private readonly isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
```
Comment (alan-agius4) [inline packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:50]: NIT: Move this into the constructor as it is not used anywhere else. Since properties are not minified, this approach will also help to optimize the code.

File: packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts [hunk 3]
```diff
@@ -98,20 +100,20 @@ export class LCPImageObserver implements OnDestroy {
...
 
   unregisterImage(rewrittenSrc: string) {
     if (!this.observer) return;
-    this.images.delete(getUrl(rewrittenSrc, this.window!).href);
+    this.images.delete(getUrl(rewrittenSrc, this.platformLocation.href).href);
   }
 
   updateImage(originalSrc: string, newSrc: string) {
-    const originalUrl = getUrl(originalSrc, this.window!).href;
+    const originalUrl = getUrl(originalSrc, this.platformLocation.href).href;
```
Comment (alan-agius4) [inline packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:112]: Inside of accessing `this.platformLocation.href` multiple times in the same method, store the value in a `const`. This will save a few bytes.

File: packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts [hunk 4]
```diff
@@ -42,14 +45,13 @@ export class LCPImageObserver implements OnDestroy {
   // Map of full image URLs -> original `ngSrc` values.
   private images = new Map<string, ObservedImageState>();
 
-  private window: Window | null = null;
   private observer: PerformanceObserver | null = null;
+  private platformLocation = inject(PlatformLocation);
```
Comment (alan-agius4) [inline packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:49]: You can also change this to the below to save some extra bytes.

```suggestion
  private baseUrl = inject(PlatformLocation).href;
```

**Review by alan-agius4 — COMMENTED** [review]
LGTM, couple of minor nits.

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:646510455:BASE:646510494:1719333790508:116ec2fe).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit is "green" for the changes in this PR, this PR is ready for merge.

**Comment by thePunderWoman** [comment]
This PR was merged into the repository by commit 39e48ce6757ac2dc4157ab69acade0ff1cd2208e.

The changes were merged into the following branches: main, 18.0.x

**Comment by AndrewKushnir** [comment]
The change will be reverted via https://github.com/angular/angular/pull/56740, since there are some breakages in Google's codebase. The error looks like this:

```
TypeError: Failed to construct 'URL': Invalid base URL

    at getUrl (packages/common/src/directives/ng_optimized_image/url.ts:12
    at LCPImageObserver.registerImage (packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:103
    at packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:438
   ...
```

I think we may need to split this PR into 2:

- one that would add a flag to stop running some code on the server (low risk)
- additional refactoring to use platform location, etc (higher risk, would require a TGP)

**Comment by JeanMeche** [comment]
Back to review. 
This is now the minimal fix by checking for the platform.

**Comment by AndrewKushnir** [comment]
> This is now the minimal fix by checking for the platform.

@JeanMeche I still see changes related to platform location in the diff. Could you please check if you've pushed the most recent version of the change?

File: packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts [hunk 5]
```diff
@@ -98,20 +97,20 @@ export class LCPImageObserver implements OnDestroy {
...
+    this.images.set(getUrl(rewrittenSrc, window).href, newObservedImageState);
   }
 
   unregisterImage(rewrittenSrc: string) {
     if (!this.observer) return;
-    this.images.delete(getUrl(rewrittenSrc, this.window!).href);
+    this.images.delete(getUrl(rewrittenSrc, window).href);
   }
 
   updateImage(originalSrc: string, newSrc: string) {
```
Comment (AndrewKushnir) [inline packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:108]: It'd be great to add an exit condition into this function as well:
```suggestion
  updateImage(originalSrc: string, newSrc: string) {
    if (!this.observer) return;
```

File: packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts [hunk 6]
```diff
@@ -42,14 +43,12 @@ export class LCPImageObserver implements OnDestroy {
...
   private images = new Map<string, ObservedImageState>();
 
-  private window: Window | null = null;
   private observer: PerformanceObserver | null = null;
 
   constructor() {
+    const isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
     assertDevMode('LCP checker');
-    const win = inject(DOCUMENT).defaultView;
-    if (typeof win !== 'undefined' && typeof PerformanceObserver !== 'undefined') {
```
Comment (AndrewKushnir) [inline packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:51]: I think it'd be great to keep the logic that uses `DOCUMENT` as is, since this token can be overridden (e.g. for testing purposes).

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche thanks for addressing the feedback 👍 I think we can simplify the change more and keep the original code that works with the `DOCUMENT` token.

File: packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts [hunk 7]
```diff
@@ -68,16 +70,10 @@ export class PreconnectLinkChecker {
    */
   private alreadySeen = new Set<string>();
 
-  private window: Window | null = null;
-
   private blocklist = new Set<string>(INTERNAL_PRECONNECT_CHECK_BLOCKLIST);
 
   constructor() {
     assertDevMode('preconnect link checker');
-    const win = this.document.defaultView;
```
Comment (AndrewKushnir) [inline packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:77]: Could you please revert these changes (similar to `packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts`) as well to minimize the diff?

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:646510455:BASE:651840610:1720809868761:339452dd).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: the presubmit is "green", this PR is ready for merge. **Important note**: this PR was previously reverted from g3 due to some breakages introduced by extra refactoring. This version of PR contains the minimum set of changes required, so the risk of rollback is low. However, please merge and sync this PR individually (preferably early in the week as well).

**Comment by atscott** [comment]
This PR was merged into the repository by commit 2c4613a002d7670f8377cb53eaa9aca4bfc9521f.

The changes were merged into the following branches: main, 18.1.x
